resharper/src/resharper-unity/CSharp/Feature/Services/QuickFixes/UseExplicitTypeInsteadOfStringQuickFix.cs

```diff
@@ -0,0 +1,138 @@
+            myStringLiteral = warning.StringLiteral;
+            myAvailableTypes = warning.AvailableTypes;
+        }
+
+        public bool IsAvailable(IUserDataHolder cache)
+        {
+            return myInvocationExpression.IsValid();
+        }
+
+
+        private IReadOnlyList<IBulbAction> GetItems()
+        {
+            return myAvailableTypes.Select(t =>
+                    new UseExplicitTypeInsteadOfStringAction(t, myInvocationExpression, myMethodName, myStringLiteral))
+                .ToList();
+        }
+
+        public IEnumerable<IntentionAction> CreateBulbItems()
+        {
+            foreach (var action in GetItems())
+            {
+                var anchor = new SubmenuAnchor(CreateFromUsageAnchor, SubmenuBehavior.Executable);
+                yield return new IntentionAction(action, null, anchor);
+            }
+        }
+
+
+        private class UseExplicitTypeInsteadOfStringAction : BulbActionBase
+        {
+            private readonly ITypeElement myType;
+            private readonly IInvocationExpression myOldInvocation;
+            private readonly string myMethodName;
+            private readonly string myLiteralName;
+
+            public UseExplicitTypeInsteadOfStringAction(ITypeElement type, IInvocationExpression oldInvocation,
+                string methodName, string literalName)
+            {
+                myType = type;
+                myOldInvocation = oldInvocation;
+                myMethodName = methodName;
+                myLiteralName = literalName;
+            }
+
+            protected override Action<ITextControl> ExecutePsiTransaction(ISolution solution,
+                IProgressIndicator progress)
+            {
+                using (WriteLockCookie.Create())
+                {
+                    var factory = CSharpElementFactory.GetInstance(myOldInvocation);
+                    if (myOldInvocation.InvokedExpression is IReferenceExpression referenceExpression)
+                    {
+                        var builder = FactoryArgumentsBuilder.Create();
+
+                        var qualifier = referenceExpression.QualifierExpression;
+                        if (qualifier != null)
+                        {
+                            builder.Argument(qualifier);
+                            builder.Append(".");
+                        }
+
+                        builder.Append(myMethodName);
+                        builder.Append("<");
+                        builder.Append(myType.GetClrName().FullName);
+                        builder.Append(">");
+                        builder.Append("()");
+
+                        var newInvocation = factory.CreateExpression(builder.ToString(), builder.ToArguments());
+                        var result = ModificationUtil.ReplaceChild(myOldInvocation, newInvocation);
+
+                        // beatify generated reference
+                        var resultTextRange = result.GetDocumentRange();
```

Comment:
This part should be redundant if you use `builder.Argument(IType);` instead of a fully qualified name.
Current implementation also unnecessarily optimizes references in qualifier which might not be expected. Might worth adding a test.

Also please check `.IsValid()` when using document ranges.